Use BDDs for encoding symex guards

romainbrenguier opened the pull request:
The main commit is "Use BDDs for encoding guards in symex". Commits that are before are in preparation for this one (in particular getting rid of usage of functions the new version will not implement). Commits that follow are clean-up in the code that use the guard (in particular getting rid of simplifications that are now unnecessary).

I still wan't to run this on more benchmarks before it is merged. For what I've seen so far, this is about 10% faster than the current symex, but this is not uniform across benchmarks (it doesn't help for "simple" instances, i.e. which take less than 2s).


<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented:
Before reviewing this in any way: I believe I had mentioned #310. Does this PR relate to it in any way?

romainbrenguier commented:
@tautschnig 
> Before reviewing this in any way: I believe I had mentioned #310. Does this PR relate to it in any way?

Yes, the idea is similar, but this is taking it one step further: instead of keeping guards as `exprt` and using BDDs to simplify them at some points in the code, we always keep the guards as BDDs and only convert to `exprt` when `as_expr` is called.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1342,8 +1342,7 @@ void goto_checkt::check_rec(const exprt &expr, guardt &guard, bool address)
 
-    guard.swap(old_guard);
-
+    guard = std::move(old_guard);
```
Almost randomly poking at commits: what's the improvement here? (I actually have a yet-to-be-PR'ed commit lying around making the opposite change, i.e., replacing uses of `std::move` by `swap`.)

romainbrenguier replied:
The new implementation of guardt doesn't have a `swap` (though exprt has). And since `old_guard` is not used afterward, I don't see a reason for using swap?

tautschnig replied:
For `irept`-based implementations `= std::move(...)` is at best as efficient as `swap`, and possibly less efficient if `USE_MOVE` is not enabled. (But I do concede that we should just get rid of `USE_MOVE` as an optional feature as we require C++11 anyway.) So in that setting the change doesn't seem to provide any added value.

With the new implementation of `guardt` not being an `irept`, the situation of course is different. But that needs to be spelled out in the commit message...

tautschnig reviewed (changes requested):
Given this is marked "In progress" I'm sure this is the plan, but just to be explicit: please turn this into a series of PRs, there are a lot of seemingly unrelated changes.

tautschnig commented:
> Yes, the idea is similar, but this is taking it one step further: instead of keeping guards as exprt and using BDDs to simplify them at some points in the code, we always keep the guards as BDDs and only convert to `exprt` when `as_expr` is called.

Ok, but do note that in terms of history this is a step backwards: `guardt` used to be unrelated to `exprt`, and making it an `exprt` enabled access to more sharing when previously we had a lot of redundant copies.

tautschnig commented on src/analyses/goto_rw.cpp:
```diff
@@ -687,15 +687,15 @@ void rw_guarded_range_set_value_sett::get_objects_if(
+    guardt copy = guard;
 
-    guard.add(not_exprt(if_expr.cond()));
+    copy.add(not_exprt(if_expr.cond()));
```
This should be `guard.add`

tautschnig commented:
There is a lot of great cleanup in this PR that should be merged as soon as possible, but via individual PRs (really one per commit). Then this PR can be reduced to the necessary minimum and also the effect of BDDs (and not that of the cleanup) can be measured.

romainbrenguier committed:
New implementation of guards using BDDs

This is not used yet in CBMC.
The aim is to specify at compile time whether we want to use this new
implementation or the old one.

Using BDDs ensures the guards are always in some "simple" form.

romainbrenguier committed:
Move guard implementation to guard_expr.h/cpp

This is to clearly seperate the two implementations.
The choice between the two will be made in guard.h

romainbrenguier committed:
Rename class to guard_exprt

To avoid confusion with the other guard implementation.

This forces us to use includes instead of forward declaration for guardt
because forward declaration doesn't work well with typedef.

romainbrenguier committed:
Switch between guard bdd and guard expr

Depending on the BDD_GUARDS flag, we will use the expr implementation or
the BDD implementation.

romainbrenguier committed:
Apply clang-format suggestions

No functional change

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 512d5b53182d9c39d2acdf37cd4985cd9cd7eb9e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97400932
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: ec61e68aa856a18d34eafd97e510d3d992a84138).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97402213
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

martin-cs commented:
I think I am with @tautschnig on this needing to be broken up to be merged.  Also, please have go with the example from https://github.com/diffblue/cbmc/issues/306 and see if it improves things there.  I believe this should make a significant performance difference so it would be good to have it profiled to check that it is a positive difference.

romainbrenguier commented:
@tautschnig @martin-cs I've extracted two PRs from this https://github.com/diffblue/cbmc/pull/3820 https://github.com/diffblue/cbmc/pull/3821 and I will rebase this one once these are merged. In the mean time I will run some benchmarks

romainbrenguier commented:
@martin-cs 
> Also, please have go with the example from #306 and see if it improves things there. 

Unfortunately on this example this makes symex about 5 times slower, but then the solver is about 5 times faster (total time is about 4 times worse). I will try to do some profiling to see where we lose time.

romainbrenguier commented:
> Unfortunately on this example this makes symex about 5 times slower, but then the solver is about 5 times faster (total time is about 4 times worse). I will try to do some profiling to see where we lose time.

I found where time was lost, it's the `do_simplify(simpl_state_guard);` in symex_goto.cpp:109, it should now be unnecessary with BDDs and if I remove it, cbmc is twice as fast as before on the example.

romainbrenguier committed:
Add is_always_simplified field in guard implementations

Expressions obtained from BDDs are already simple. Trying to simplify
them is redundant and can slow down the execution.
We add this field so that symex code can take the decision to simplify
or not an expression based on this information.

romainbrenguier commented:
This is now based on https://github.com/diffblue/cbmc/pull/3820 which should be merged first

romainbrenguier commented:
@tautschnig This is rebased and now ready for review

romainbrenguier commented:
Here are the results for some benchmarks. On average the BDD version about 17% faster, and on some benchmarks it was 65% faster. So I think this has a good potential.
![perf_bdd](https://user-images.githubusercontent.com/11854194/51475789-5c67cc00-1d7b-11e9-993b-ba7725437d3e.png)

tautschnig commented:
What would be the effect of using BDDs to simplify Boolean expressions rather than for managing guards in general? Would that yield more or less the same benefits? I'm asking as this might result in smaller, more local changes.

peterschrammel commented on src/util/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 file(GLOB_RECURSE sources "*.cpp" "*.h")
 
+
```
🚫 Why is a java-models-library bump in this commit c04d5960b?

romainbrenguier replied:
that's a mistake

peterschrammel commented on src/util/bdd.h:
```diff
@@ -0,0 +1,146 @@
+/// the BDD functions should only be accessed from this header file.
+
+#ifndef CPROVER_UTIL_BDD_H
+#define CPROVER_UTIL_BDD_H
```
❓ 💡 🚫 This should rather be put this into `solvers/bdd`. However, this would require `guardt` (currently in `util`) to be moved somewhere else. `guardt` is used by `goto-symex` and `analyses`. 
I see the following options: Move `guardt` into
- `solvers/guard`
- `analyses`
- new module?

romainbrenguier replied:
I'm not sure BDDs should go to `solvers` because they can do more than solving system of equations. They can be use as an alternative to exprt for representing Boolean expressions, so I find `util` appropriate.

peterschrammel replied:
If you place it in `util` you cannot provide an implementation using miniBDD without moving miniBDD into `util` or creating a new module containing BDD things on which `util` depends .

peterschrammel commented on src/util/guard_bdd.h:
```diff
@@ -0,0 +1,80 @@
+#include "make_unique.h"
+#include "std_expr.h"
+
+struct guard_bdd_managert
```
❓ 🚫  Who owns this manager object?

romainbrenguier replied:
There are several ones: in goto_checkt, goto_symext, rd_range_domaint, dep_graph_domaint, and goto_instrument_parse_options

kroening commented:
BTW, did you try miniBDD?
cudd is a non-trivial dependency.

romainbrenguier commented:
@kroening 
> BTW, did you try miniBDD?
> cudd is a non-trivial dependency.

It's not a strict dependency, so cbmc will continue to compile without downloading cudd, unless you add `-DCMAKE_USE_CUDD=true` to the cmake configuration. I could add a second implementation of `bdd.h` using miniBDD in case we want to use BDDs without downloading cudd, but the fact that it has no equivalent of `Constrain` makes me believe it would be slower than the implementation without BDDs in a lot of cases.

tautschnig commented:
@kroening @romainbrenguier Would it be worth implementing `Constrain` in `miniBDD`? CUDD scares me, because 1) it hardly seems maintained and 2) it's large. This "I'm scared" is amplified by the necessary intrusiveness of the changes in this PR, which means we also cannot easily get rid of this again.

romainbrenguier commented:
@tautschnig 
> This "I'm scared" is amplified by the necessary intrusiveness of the changes in this PR, which means we also cannot easily get rid of this again.

I don't see what you mean, you can remove the file `bdd.h`, `guard_bdd.h/cpp` and everything will compile as before.

romainbrenguier commented:
Here are the performance I got with miniBDD when compared with develop:
![perf_minibdd](https://user-images.githubusercontent.com/11854194/51684559-385af380-1fe4-11e9-9991-1db35b9ac867.png)

It is sometimes faster but globally slower than develop. I guess implementing Constrain could solve that.

romainbrenguier commented:
I tried with my own implementation of constrain and it is slightly better but still not as good as without BDDs:
![perf_minibdd_constr](https://user-images.githubusercontent.com/11854194/51684616-588ab280-1fe4-11e9-9ad6-62403a0eb9ea.png)

romainbrenguier commented:
The implementation of the miniBDD interface https://github.com/romainbrenguier/cbmc/commit/03268e03de906338d225b46f8a77fef63aa79e60
and of constrain https://github.com/romainbrenguier/cbmc/commit/7af788393e7b146349fc4c3e8383cc8ca6076394
in case you are interested.

romainbrenguier committed:
Move guardt to analyses module

We want to add an implementation of guard that uses BDDs, which are part
of the solvers module. We cannot add a dependency of the util module on
solvers.

romainbrenguier committed:
Add guard_manager class and pass it around

This in preparation to using BDDs to encode guards,
every part of the code creating guards need to reference a guard_manager.

romainbrenguier committed:
Reformat guard.cpp and guard.h

No functional changes.
This was done automatically using clang-format.
This is in preparation to moving this two files, so that clang-format
doesn't reformat the file at the same time they are moved, and so that
the move commit is clean.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 2cf73c7425ca0ea5011b604f0bca8fa603faeb61).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100133143
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8a9a16ec6e3862a7532c48af9a4c2b28eb2eea97).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100212188

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c1eb0916434c4d69b56776f191fb6534ddf8d2af).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100598628

romainbrenguier commented:
@tautschnig @peterschrammel this is rebased and passing CI

tautschnig commented on src/analyses/guard_bdd.cpp:
```diff
@@ -0,0 +1,306 @@
+}
+
+/// Convert an expression to a bddt
+static bddt bdd_from_expr(const exprt &expr, guard_bdd_managert &manager)
```
How does this relate/compare to `solvers/prop/bdd_expr.{h,cpp}`?

romainbrenguier replied:
Indeed there are very similar. Originally the bdd_expr implementation was for miniBDD and this one for Cudd but now that they are both based on the unified interface the `from_expr` and `as_expr` are basically duplicates. ~I could move the bdd_expr from solvers/prop to analyses and use that in guard_bdd,~ but I think the class needs to be adapted a bit, in particular the `bdd_exprt` shouldn't own the bdd_manager and the expr to bdd map.

romainbrenguier replied:
sorry for the confusion I, don't need to move bdd_expr since analyses can use solvers. I will just add to the interface of `bdd_expr` to expose `from_expr_rec` and `as_expr` as non member functions, and use these directly from `bdd_guard`, no need for changing the content of the `bdd_exprt` class

tautschnig replied:
Ok, whatever helps to avoid having two almost-the-same implementations seemingly solving the same problem. I don't mind you replacing my past implementation, I just want to make sure we don't have duplicate code.

tautschnig commented on src/util/Makefile:
```diff
@@ -26,7 +26,6 @@ SRC = allocate_objects.cpp \
       fresh_symbol.cpp \
       get_base_name.cpp \
       get_module.cpp \
-      guard.cpp \
```
Wrong commit?

tautschnig commented on src/analyses/dependence_graph.h:
```diff
@@ -197,6 +197,9 @@ class dep_graph_domaint:public ai_domain_baset
   // location has a data dependency on
   depst data_deps;
 
+  // Used for
```
for ... what?

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -85,12 +85,13 @@ class goto_checkt
     guardt &guard,
     bool address);
-  void check(const exprt &expr);
+  void check(const exprt &expr, guard_managert &guard_manager);
```
Why both pass it around and own it?

tautschnig commented on src/analyses/dependence_graph.h:
```diff
@@ -213,7 +216,8 @@ class dep_graph_domaint:public ai_domain_baset
     dependence_grapht &dep_graph,
-    const namespacet &ns);
+    const namespacet &ns,
+    guard_managert &guard_manager);
```
Both passing it around and owning it?

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1931,9 +1932,7 @@ void goto_check(
-void goto_check(
-  const optionst &options,
-  goto_modelt &goto_model)
+void goto_check(const optionst &options, goto_modelt &goto_model)
```
No change?

romainbrenguier replied:
Yes sorry (some reformatting got mixed in, because of doing some changes, reverting them, and applying clang-format meaning the commit and its revert don't cancel each other)

tautschnig commented on src/analyses/goto_check.h:
```diff
@@ -29,9 +30,7 @@ void goto_check(
-void goto_check(
-  const optionst &options,
-  goto_modelt &goto_model);
+void goto_check(const optionst &options, goto_modelt &goto_model);
```
No change?

tautschnig commented on src/analyses/goto_rw.h:
```diff
@@ -165,6 +165,7 @@ class rw_range_sett
 
 protected:
   const namespacet &ns;
+  guard_managert &guard_manager;
```
Is it both being passed around and owned?

romainbrenguier replied:
I removed the unnecessary passing around

tautschnig commented on src/goto-instrument/accelerate/disjunctive_polynomial_acceleration.cpp:
```diff
@@ -399,13 +403,14 @@ bool disjunctive_polynomial_accelerationt::find_path(patht &path)
   polynomialt &polynomial,
-  patht &path)
+  patht &path,
+  guard_managert &guard_manager)
```
It may be better to make this a class member somewhere rather than passing it around all across the acceleration code?

romainbrenguier replied:
For readability or performance?

tautschnig replied:
For readability and impact of this PR.

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -141,6 +146,8 @@ class goto_symex_statet final : public goto_statet
   /// for error traces even after symbolic execution has finished.
   symbol_tablet symbol_table;
 
+  // Manager is requiered to be able to resize the thread vector
```
s/requiered/required/

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -141,6 +146,8 @@ class goto_symex_statet final : public goto_statet
   symbol_tablet symbol_table;
 
+  // Manager is requiered to be able to resize the thread vector
+  guard_managert guard_manager;
```
We is it also being passed to all the `rename` functions when we have a guard manager here?

tautschnig reviewed (commented):
1) I get the impression that a guard manager is being both passed around and owned by a class as the same time, which seems unnecessary. If it is unnecessary, then the churn in the last commit can likely be reduced a lot. Also there is a question about `bdd_exprt` vs the code newly introduced here.

2) I don't see how this uses the use-MiniBDD-or-CUDD alternative.

Overall, this is a very intrusive change, adding additional dependencies across directories. So let's try to keep it as clean as we possibly can.

tautschnig commented:
The merge of #4097 just now results in some need for rebasing, but really should help in that it reduces the usage of `guardt`.

romainbrenguier committed:
Add a cache to as_expr function for BDDs

Caching the intermediary results can lead to big performance
improvements since the number of calls without cache could be
potentially exponential in the size of the BDD.

romainbrenguier committed:
Document the BDD_GUARDS compilation flag

This adds instruction to COMPILING.md so that people interested in
performances can find about it, and try it.

romainbrenguier commented:
@tautschnig 
> 2\. I don't see how this uses the use-MiniBDD-or-CUDD alternative.

Yes this was a bit confusing because reusing the HAVE_CUDD flag. I've now added a new flag, BDD_GUARDS and documented it in COMPILING.md, it should be clearer and allow to use bdd guards with miniBDD.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 766baae2fdc605e9d4da7b172c3db175ee158c7a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100765659

romainbrenguier commented:
@tautschnig @peterschrammel this is ready for review

tautschnig commented on src/pointer-analysis/value_set_dereference.cpp:
```diff
@@ -16,6 +16,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/format_expr.h>
 #endif
 
+#include <analyses/guard.h>
```
The use of `guardt` in this file is pointless, could you please clean that up so that we don't even introduce a dependency on guardt (let alone BDDs) here?

tautschnig replied:
I don't think the include is needed anymore?

tautschnig commented on src/pointer-analysis/goto_program_dereference.cpp:
```diff
@@ -11,13 +11,13 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "goto_program_dereference.h"
 
+#include <analyses/guard.h>
```
I think we could safely use `exprt` instead of `guardt` here, if it's actually needed at all. We don't even try to evaluate it.

tautschnig replied:
This include should no longer be needed.

tautschnig commented on src/goto-programs/goto_convert_class.h:
```diff
@@ -16,8 +16,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <vector>
 #include <unordered_set>
 
+#include <analyses/guard.h>
```
Doesn't even use `guardt` (anymore).

tautschnig replied:
Still applies.

tautschnig commented on src/goto-instrument/rw_set.h:
```diff
@@ -18,7 +18,7 @@ Date: February 2006
 #include <set>
 
-#include <util/guard.h>
+#include <analyses/guard.h>
```
This code doesn't make meaningful use of a "guard". Should just be removed.

tautschnig commented on src/analyses/goto_rw.h:
```diff
@@ -17,7 +17,7 @@ Date: April 2010
 #include <memory> // unique_ptr
 
-#include <util/guard.h>
+#include <analyses/guard.h>
```
Should now use `#include "guard.h"`

tautschnig replied:
Still applies.

tautschnig commented on src/goto-instrument/wmm/module_dependencies.txt:
```diff
@@ -1,3 +1,4 @@
+analyses
```
I believe this won't be necessary after the cleanup.

tautschnig replied:
Still applies: this should not be necessary.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -13,6 +13,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <algorithm>
 
+#include <analyses/guard.h>
```
Should be `#include "guard.h"`

tautschnig replied:
Still applies.

tautschnig reviewed (commented):
I think we can vastly reduce the impact of the changes by trimming down the use of `guardt` to where it really is used. I have only looked at the first commit as some of the others ought to become a lot simpler once the cleanup is done. We're turning `guardt` into a much more powerful tool, and we should make sure we only use it where we benefit from (or even make any use of at all!) the value this adds over a plain Boolean `exprt`.

romainbrenguier commented:
> I think we can vastly reduce the impact of the changes by trimming down the use of guardt to where it really is used. I have only looked at the first commit as some of the others ought to become a lot simpler once the cleanup is done. We're turning guardt into a much more powerful tool, and we should make sure we only use it where we benefit from (or even make any use of at all!) the value this adds over a plain Boolean exprt.

Agreed. I will do some clean-up.

romainbrenguier commented:
@tautschnig I opened a new PR for the clean-up https://github.com/diffblue/cbmc/pull/4188 (in order to not make this one too complex)

romainbrenguier commented:
@tautschnig I rebased, making the big commit smaller. It seems guards are only used in `goto-symex` and `goto-instrument/accelerate` now.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: aa6017f39b4d22300de092ed493c65554cfed679).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101306605

romainbrenguier commented:
@tautschnig @peterschrammel this is ready for review

tautschnig commented on src/analyses/guard.h:
```diff
@@ -50,6 +64,8 @@ class guardt
   friend guardt &operator-=(guardt &g1, const guardt &g2);
   friend guardt &operator|=(guardt &g1, const guardt &g2);
 
+  guard_managert &manager;
```
Should this really be `public`?

romainbrenguier replied:
It doesn't seem so. I will fix that.

tautschnig replied:
Though see also my later comment: maybe you can remove that member completely? It has no use in the `guardt`/`guard_exprt` implementation.

romainbrenguier replied:
I removed it from the `guard_exprt` implementation.

tautschnig commented on src/analyses/guard.h:
```diff
@@ -16,13 +16,27 @@ Author: Daniel Kroening, kroening@kroening.com
+  guardt &operator=(const guardt &other)
+  {
+    expr = other.expr;
+    manager = other.manager;
```
Isn't that possibly expensive? I am mainly worried about having a non-moving assignment operator as all other implementations (such as BDD-backed guards) will need to provide the same.

romainbrenguier replied:
This is just a reference so it should be fine.

tautschnig replied:
I am worrying about the case where `manager` includes some data structures like maps. Doing a non-moving assignment is pretty expensive in that case. I acknowledge that the BDD-backed implementation actually doesn't do any copying. Maybe all you should do is remove the class member from the non-BDD guard? You can still consume a manager in the constructor, just don't use it...

romainbrenguier replied:
Oh yes you are right, we don't need a manager field for guard_expr

tautschnig commented on src/goto-instrument/accelerate/accelerate.cpp:
```diff
@@ -85,7 +85,9 @@ bool acceleratet::contains_nested_loops(goto_programt::targett &loop_header)
-int acceleratet::accelerate_loop(goto_programt::targett &loop_header)
+int acceleratet::accelerate_loop(
+  goto_programt::targett &loop_header,
+  guard_managert &guard_manager)
```
Do we really need to pass it around everywhere? Can't we make it a class member instead? Lots of churn here.

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -190,6 +190,8 @@ class goto_symext
   /// goto-program, and the names of dynamically-created objects.
   namespacet ns;
+
+  guard_managert guard_manager;
```
I thus far know very little about what a `guard_managert` is, and in particular when it can/cannot be shared or maybe must be shared. Here, I am wondering whether `goto_symext` is the right place, or should it rather be `path_storaget`?

romainbrenguier replied:
Guard managers are used to create guards, and the important thing is that guards created with different guard managers should not mix together. I think guards created by goto_symext can never exit its scope, so it should be fine. Maybe `path_storaget` would be better, but I don't know what is the relationship between the two.

tautschnig replied:
The `path_storaget` will live across several `goto_symext` instantiations. But you're certainly right, those guards won't mix.

Maybe you can put a comment somewhere saying what you said here about the constraints on guards/guard managers?

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -74,7 +74,11 @@ class goto_statet
   explicit goto_statet(const class goto_symex_statet &s);
-  explicit goto_statet(const symex_targett::sourcet &_source) : source(_source)
+
+  explicit goto_statet(
```
`explicit` no longer needed.

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -127,14 +131,16 @@ struct framet
 {
 public:
-  explicit goto_symex_statet(const symex_targett::sourcet &);
+  explicit goto_symex_statet(
```
`explicit` no longer needed.

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -74,7 +74,7 @@ void goto_symext::symex_assign(
       assignment_type=symex_targett::assignment_typet::HIDDEN;
 
-    guardt guard{true_exprt{}}; // NOT the state guard!
+    guardt guard{true_exprt{}, guard_manager}; // NOT the state guard!
```
I am wondering whether there is actual value in using `guardt` for _this_ case. It's just there to bring if-then-else expressions from the left-hand side to the right-hand side.

romainbrenguier replied:
It looks like it could be replaced by a conjunction, the only operation we do on them is adding more conjuncts. I can make this type to `exprt::operandst` and rename it `guard_conjuncts` if you think that's a good idea.

tautschnig replied:
That would be nice, thank you!

tautschnig replied:
And while at it, avoid use the term "guard" in this? Maybe `lhs_if_then_else_conditions`? And then of course kill that comment ("NOT the state guard"). It's not a guard, it shouldn't be named "guard", and thus the confusion is gone.

romainbrenguier replied:
`lhs_if_then_else_conditions` is not very clear to me, maybe because I don't really understand what this represents. It seems like it's conditions that have to hold for the assignment to take place, in that case it could be named `assignment_conditions`?

tautschnig replied:
I'm not sure about the best name either. Yes, it is a list of conditions for an assignment to use a particular symbol on the LHS. An example: `(cond ? foo : bar) = 42` will yield:
```
foo = 42  [assignment condition: cond]
bar = 42  [assignment condition: !cond]
```
(FWIW: those if-then-else on the left-hand side are a result of dereferencing)

tautschnig commented on src/goto-symex/symex_main.cpp:
```diff
@@ -338,7 +339,7 @@ void goto_symext::initialize_path_storage_from_entry_point_of(
   auto state = initialize_entry_point_state(get_goto_function);
 
-  path_storaget::patht entry_point_start(target, *state);
+  path_storaget::patht entry_point_start{target, *state};
```
Unnecessary change.

tautschnig reviewed (commented):
Only reviewed up to "Make from_expr_rec a non-member function", will continue a bit later.

tautschnig commented on src/solvers/prop/bdd_expr.cpp:
```diff
@@ -57,30 +65,32 @@ bddt bdd_exprt::from_expr_rec(const exprt &expr)
-      node_map.push_back(expr);
-      const unsigned index = (unsigned)node_map.size() - 1;
+      var_to_expr.push_back(expr);
+      const unsigned index = (unsigned)var_to_expr.size() - 1;
```
As you are changing this line, could you please introduce `narrow_cast` here?

tautschnig commented on src/solvers/prop/bdd_expr.h:
```diff
@@ -27,6 +27,13 @@ Author: Michael Tautschnig, michael.tautschnig@qmul.ac.uk
+  const exprt &expr,
+  bdd_managert &bdd_mgr,
+  std::unordered_map<exprt, bddt, irep_hash> &expr_to_var,
+  std::vector<exprt> &var_to_expr);
```
Despite the commit message I still don't understand why this is necessary or even the right thing to do. Specifically, it exposes implementation details (the map and the vector) that need to be in sync. The purpose of the `bdd_exprt` class was to encapsulate these implementation details.

romainbrenguier replied:
Yes, you are right this should be encapsulated. My problem with `bdd_exprt` is that it binds together the `bddt` and the `bdd_managert` meaning get only one `bddt` by manager, whereas the same manager should be shared between many BDDs. If that's fine, I will change the interface of `bdd_exprt` so that the result of the conversion is returned instead of being stored in a field.

tautschnig replied:
That would seem like a much better approach. Feel free to fiddle with the `bdd_managert` member (for example, it might become a reference member). I'm not picky about the final `bdd_exprt` interface, but I'd like to maintain appropriate encapsulation. Data structures that need to be kept in sync are bound to get us in trouble, and thus must be hidden behind an interface.

tautschnig commented on src/solvers/prop/bdd_expr.h:
```diff
@@ -34,6 +34,9 @@ bddt bdd_from_expr(
   std::vector<exprt> &var_to_expr);
 
+/// Convert a BDD node to an expression.
+exprt bdd_as_expr(const bdd_nodet &node, const std::vector<exprt> &var_to_expr);
```
Just as with `bdd_from_expr`: I still don't understand why this is the right thing to do. It exposes an implementation detail.

tautschnig commented on src/solvers/prop/bdd_expr.cpp:
```diff
@@ -109,10 +109,12 @@ void bdd_exprt::from_expr(const exprt &expr)
   const std::vector<exprt> &var_to_expr,
-  bool complement)
+  bool complement,
+  std::unordered_map<long, exprt> &cache)
```
Why is `long` the right data type? I think I never got an answer on the other BDD PR why the `id()` function has that type. I am still concerned about it.

romainbrenguier replied:
The right type for Cudd would be something like `DdNode*`, and for other implementations of BDDs this type does not exists. Using `long` is a way of making the type independent of the implementation, though having to cast a pointer to `long` is not nice. Any better idea?

tautschnig replied:
Can we use a typedef (as a member of the `bdd_nodet` classes)? Then you would just use `bdd_nodet::id_type` (or whatever the name). With the current set-up we do not require that `bdd_nodet::id()` returns that same type in all implementations, just some type that is usable, e.g., as a key in a map.

(Also, `long` would not be sufficient to store a pointer on 64-bit Windows.)

romainbrenguier replied:
good id

tautschnig commented on src/analyses/guard_bdd.h:
```diff
@@ -0,0 +1,81 @@
+#include <util/make_unique.h>
+#include <util/std_expr.h>
+
+struct guard_bdd_managert
```
I note that this is entirely unrelated to `guard_managert`. Shouldn't they at least share an interface?

tautschnig commented on src/goto-instrument/mmio.h:
```diff
@@ -14,6 +14,8 @@ Date: September 2011
 #ifndef CPROVER_GOTO_INSTRUMENT_MMIO_H
 #define CPROVER_GOTO_INSTRUMENT_MMIO_H
 
+#include <analyses/guard.h>
```
That's surprising?!

romainbrenguier replied:
This has been left from old commits, which should have been cleaned-up after rebase. I'll remove them.

tautschnig commented on src/goto-instrument/wmm/goto2graph.h:
```diff
@@ -20,6 +20,7 @@ Date: 2012
 #include <util/namespace.h>
 #include <util/message.h>
 
+#include <analyses/guard.h>
```
More surprises.

tautschnig commented on src/analyses/guard.h:
```diff
@@ -12,9 +12,20 @@ Author: Daniel Kroening, kroening@kroening.com
+#include "guard_bdd.h"
+
+using guard_managert = guard_bdd_managert;
+using guardt = guard_bddt;
```
Would it be possible for them to please have a shared interface? Otherwise inconsistencies will creep in that can only be detected by compiling both with and without `BDD_GUARDS` set.

romainbrenguier replied:
I'm reluctant to use virtual functions since they are not really needed here. I could try doing that using CRTP if you think that's a good idea (`guard_bdd_managert` would inherit from `guard_interfacet<guard_bdd_managert>`).

tautschnig replied:
I suppose we had the same discussion on the CUDD/miniBDD BDD PR :-)  It's not the most important of all issues and can be left to a future PR - but I will remind you about it whenever there is a change to these interfaces and someone forgets to update both of them :-P

romainbrenguier replied:
fair enough

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -108,7 +108,10 @@ void goto_symext::symex_goto(statet &state)
+
+#ifndef BDD_GUARDS
   do_simplify(simpl_state_guard);
+#endif
```
I'd recommend changing the `.as_expr()` interface to take a Boolean argument whether or not to run the simplifier (inside `.as_expr`). It could have a default value (`false`) so as not to break existing APIs.

romainbrenguier replied:
I don't like adding arguments that would always be called with a constant. An alternative would be to define a const static field of the guardt class that tells us whether guards benefit from simplifications.  That way an `if` condition on this Boolean should even get simplified by the compiler.

tautschnig replied:
Sounds like a good approach to me!

tautschnig commented on COMPILING.md:
```diff
@@ -309,3 +309,22 @@ If compiling with cmake:
+There are two implementation for symex guards. The default one uses the
+internal representation of expression. The other one uses BDDs and
+though experimental, it is expected to have better performances on
+difficult programs, in particular when used in conjunction with CUDD.
```
I don't know what a "difficult" program is - maybe just omit that? I think you've demonstrated performance improvements across the board for any runtime greater than a few seconds.

tautschnig commented on COMPILING.md:
```diff
@@ -309,3 +309,22 @@ If compiling with cmake:
+
+There are two implementation for symex guards. The default one uses the
+internal representation of expression. The other one uses BDDs and
+though experimental, it is expected to have better performances on
```
"performance" (this is a non-countable use)

tautschnig commented on COMPILING.md:
````diff
@@ -309,3 +309,22 @@ If compiling with cmake:
+compilation flag:
+  * If compiling with make:
+    ```
+    make -C src CXX_FLAGS="-DBDD_GUARDS"
````
`CXXFLAGS` (no underscore) - makes me wonder: is this actually tested? And you will want to recommend something like `CXXFLAGS="-O2 -DBDD_GUARDS"` as otherwise the performance will be rather terrible...

romainbrenguier replied:
Corrected. I added a travis step with the flag, But it is using cmake. Is it worth adding a step with make and BDD_GUARDS?

tautschnig replied:
I think it's ok if you just test it at least once locally. I don't really expect random breakage of the Makefile-based builds with this flag. Alternatively, you could just add the flag to one of the existing Makefile-based build configurations.

romainbrenguier committed:
Remove unused field ns field of bdd_exprt

A reference to a namespace was stored without being ever used.

romainbrenguier committed:
Remove root field of bdd_exprt

Instead the result of from_exprt and the input of as_expr should be
BDDs.
This makes it possible to reuse the same manager for several exprt
conversion and to combine the results obtain from the from_expr
conversion with BDD operations.

romainbrenguier committed:
Define idt type for bdd_node::id() functions

The type of identifier can vary depending on the BDD implementation
used. Using long there was kind of a hack and not correct on
architecture where pointers wouldn't fit in a long.

romainbrenguier committed:
Enable BDD_GUARDS in one of the travis steps

We add the BDD_GUARDS flags in the travis step which is compiling with
CUDD to test that the bdd_guard implementation.

romainbrenguier committed:
Replace use of guardt by operandst in symex_assign

The guard used there was actually only used as a conjunction, so it can
be replaced by a list of expressions, of which we take the conjunction
when needed.

tautschnig commented:
@romainbrenguier clang-format and cpplint are not quite happy yet. (The other Travis failure is spurious and will go away once you force-push with the whitespace problems fixed.)

romainbrenguier committed:
Deactivate a test which takes too long with BDDs

One of the JBMC tests will take a long time when BDD are used.
We deactivate it for now in the case BDD_GUARDS is defined.
This is acceptable since using BDDs for guards is still experimental.
Ideally we should find what takes too long in the execution and optimize
it, but it may be that we hit a particular class of guards for which the
BDDs have to be big.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d73a1a5df61b2965dee3020291aa503a08a840af).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102494282
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

romainbrenguier commented:
@tautschnig @peterschrammel I had to flag a test because it was taking too long in the build with BDDs. I also had to move the guard_manager instantiation a bit higher in the hierarchy because it seems guards can be reused from one goto_symext to another (you can check the fixup commits for that change)

chrisr-diffblue commented on src/analyses/guard_bdd.h:
```diff
@@ -35,6 +35,11 @@ class guard_bddt
   guard_bddt &append(const guard_bddt &guard);
   exprt as_expr() const;
 
+  /// Tells wether the result of \ref as_expr is always in a simplified form
```
wether -> whether

chrisr-diffblue commented on src/analyses/guard_expr.h:
```diff
@@ -50,6 +50,11 @@ class guard_exprt
     return expr;
   }
 
+  /// Tells wether the result of \ref as_expr is always in a simplified form
```
wether -> whether

chrisr-diffblue reviewed (approved):
For the areas I'm a code owner for, this seems ok.

tautschnig commented on jbmc/regression/jbmc/CMakeLists.txt:
```diff
@@ -2,9 +2,18 @@ add_test_pl_tests(
+            "-C;-X;symex-driven-lazy-loading-expected-failure;-X;bdd-expected-timeout;-s;symex-driven-loading"
+            "CORE"
+    )
+endif()
```
GitHub complains about no-newline-at-EOF

tautschnig commented on jbmc/regression/jbmc/class-literals/test.desc:
```diff
@@ -1,4 +1,4 @@
-CORE
+CORE bdd-expected-timeout
```
Could you please include a comment at the bottom of this file what the behaviour with BDDs is? Is it taking minutes, hours, days? And maybe also an idea why that's happening on this particular test?

tautschnig commented on jbmc/regression/jbmc/CMakeLists.txt:
```diff
@@ -2,9 +2,18 @@ add_test_pl_tests(
+    add_test_pl_profile(
+            "jbmc-symex-driven-lazy-loading"
+            "$<TARGET_FILE:jbmc> --validate-goto-model --validate-ssa-equation --symex-driven-lazy-loading"
+            "-C;-X;symex-driven-lazy-loading-expected-failure;-X;bdd-expected-timeout;-s;symex-driven-loading"
```
We will need some equivalent thereof in the Makefile. Something like `ifeq ($(filter %DBDD_GUARDS, $(CXXFLAGS)),)`

tautschnig commented on src/analyses/Makefile:
```diff
@@ -12,6 +12,8 @@ SRC = ai.cpp \
       goto_check.cpp \
       goto_rw.cpp \
+      guard_expr.cpp \
+      guard_bdd.cpp \
```
I think "bdd" should appear before "expr", at least according to my knowledge of the alphabet ;-)

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -13,6 +13,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <algorithm>
 
+#include "guard.h"
```
Please place this with the other current-directory includes, i.e., just above "local_bitvector_analysis.h."

tautschnig commented on src/analyses/guard_bdd.cpp:
```diff
@@ -0,0 +1,96 @@
+  return *this;
+}
+
+guard_bddt &guard_bddt::operator=(guard_bddt &&other) noexcept
```
Is `noexcept` in some way required? I'm not strictly against it, but it appearing exactly once is a big surprise. I would suggest to do one of (with no particular preference):

1) Drop it.
2) Add it everywhere.
3) If it is only needed here, add a comment to clarify.

tautschnig commented on src/analyses/guard_bdd.cpp:
```diff
@@ -0,0 +1,96 @@
+  return *this;
+}
+
+void guard_bddt::guard_expr(exprt &dest) const
```
Can I just say that I never liked this interface? (I know it's not your invention.) If now is a time to fix it, would you mind doing so? I really think this function should return by value. Whether it takes a const reference or also a value I don't mind.

romainbrenguier replied:
I also didn't like it. This is now fixed.

tautschnig commented on src/analyses/guard_bdd.cpp:
```diff
@@ -0,0 +1,96 @@
+      implies_exprt tmp;
+      tmp.op0() = as_expr();
+      tmp.op1().swap(dest);
+      dest.swap(tmp);
```
Imagine a world where we just did `return implies_exprt(as_expr(), dest);` ...

romainbrenguier replied:
done

tautschnig commented on src/analyses/guard_bdd.h:
```diff
@@ -0,0 +1,77 @@
+  /// This can vary according to the guard implementation.
+  static constexpr bool is_always_simplified = true;
+
+  /// Assign dest to `guard => dest` unless guard or dest are trivial.
```
My updated comment would then be: "Return `guard => dest` or a simplified variant thereof if either `guard` or \p dest are trivial."

tautschnig commented on src/analyses/guard_expr.h:
```diff
@@ -0,0 +1,77 @@
+
+  /// Tells whether the result of \ref as_expr is always in a simplified form
+  /// and thus requires no further simplification.
+  /// This can vary according to the guard implementation.
```
I think the comment should specifically speak to this implementation.

tautschnig commented on src/goto-instrument/accelerate/scratch_program.h:
```diff
@@ -74,6 +74,7 @@ class scratch_programt:public goto_programt
   bool constant_propagation;
 
 protected:
+  guard_managert guard_manager;
```
Why both a class member and passed to individual methods?

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -108,7 +108,9 @@ void goto_symext::symex_goto(statet &state)
-  do_simplify(simpl_state_guard);
+
+  if(!guardt::is_always_simplified)
+    do_simplify(simpl_state_guard);
```
That's now gone.

tautschnig reviewed (approved):
Some mostly smallish comments below and the acceleration code still looks suspicious, but nothing blocking.

romainbrenguier committed:
Make guard_expr return a value

This avoids having an argument that is both input and output and makes
using this function nicer.

romainbrenguier commented:
@tautschnig you may want to check the last commit

tautschnig commented on src/analyses/guard_bdd.cpp:
```diff
@@ -44,24 +44,22 @@ guard_bddt &guard_bddt::operator=(guard_bddt &&other)
-      tmp.op0() = as_expr();
-      tmp.op1().swap(dest);
-      dest.swap(tmp);
+      return implies_exprt{as_expr(), expr};
```
`std::move(expr)` since it's consumed by value.

And I would get rid of a lot of braces now that the branches are one-liners.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a4cf18bab190657279757c196ac07e6e3aeb4be0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102852559
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 76a4fbe3e371e34b78dbaf05edf177384c91eace).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102854338
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c587b5ecc935679d70419c2d77479856851b955e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102856366
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 836be4497fb127b3008775444b1fde0977b30d8a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102864095
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 468c01d6555259af97129b1050cb87a0f64ba7ca).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103158445
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

tautschnig commented on jbmc/regression/jbmc/class-literals/test.desc:
```diff
@@ -1,8 +1,14 @@
+This test is deactivated for the build with BDDs because it takes more
+than 10 minutes for the test to complete.
+This is due to the size of the BDD representing the guards growing more than
+they do when reprented as exprt.
```
Nit pick: s/reprented/represented/

tautschnig commented on jbmc/regression/jbmc/class-literals/test.desc:
```diff
@@ -1,8 +1,14 @@
+than 10 minutes for the test to complete.
+This is due to the size of the BDD representing the guards growing more than
+they do when reprented as exprt.
+
```
Nit pick 2: No need for blank line

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: ddf04915ab88a0b1e7f75a8bdf0375d6972bf106).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103176652
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 22ed54bd4a508e50c1109b5298540813d82f3965).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103523152
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.
